### PR TITLE
Fix #8316: run `fix-whitespace --check` as CI workflow

### LINKED ISSUE #8316: Add whitespace check to CI
Lifted from https://github.com/haskell/cabal/issues/6370#issuecomment-1173024446 :
I think whitespace checks should be part of the CI, see e.g. https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/src/github/workflows/whitespace.yml
(configuration at https://github.com/agda/agda/blob/3044510c8966a1876951e807f1393bbc3b5e6169/fix-whitespace.yaml).

The cited workflow from the Agda repo can be copied over, with some modification to the `fix-whitespace.yaml` configuration file.

### PR DESCRIPTION
**Pull request by andreasabel**
Fix #8316: run `fix-whitespace --check` as CI workflow.

Also: fix existing whitespace violations (separate commit).

The whitespace checker can be configured in `fix-whitespace.yaml`.

### CODE REVIEW

**Comment by Mikolaj** [comment]
Would it make sense to add a line to CONTRIBUTING explaining how to run the check locally?

File: .github/workflows/whitespace.yml [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+  push:
+    branches:
+      - master
+  pull_request:
+  release:
+    types:
+      - created
+jobs:
+  check:
+    runs-on: ubuntu-20.04
```
Comment (ulysses4ever) [inline .github/workflows/whitespace.yml:13]: I think `latest` is considered preferable these days?

File: .github/workflows/whitespace.yml [hunk 2]
```diff
@@ -0,0 +1,53 @@
+name: Whitespace
+
+on:
+  push:
+    branches:
+      - master
+  pull_request:
+  release:
```
Comment (ulysses4ever) [inline .github/workflows/whitespace.yml:8]: No idea if this makes sense. Do you?

File: .github/workflows/whitespace.yml [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      - created
+jobs:
+  check:
+    runs-on: ubuntu-20.04
+
+    strategy:
+      matrix:
+        ghc-ver: ['9.2']
+        cabal-ver: [latest]
+        # stack-ver: [2.5.1]
```
Comment (ulysses4ever) [inline .github/workflows/whitespace.yml:19]: Maybe remove all mentions of stack here?

**Review by ulysses4ever — APPROVED** [review]
Lgtm modulo minor comments

File: .github/workflows/whitespace.yml (hunk 2 above)
Comment (andreasabel) [reply .github/workflows/whitespace.yml:8]: I copied this from other workflows, but don't know if we need the `release`...

**Comment by andreasabel** [comment]
Weird, how come my latest push of workflow `whitespace.yml` checked out latest `master` and thus found a whitespace violation that wasn't on the branch of this PR??
https://github.com/haskell/cabal/runs/7515526901?check_suite_focus=true#step:7:7
> Run $HOME/.local/bin/fix-whitespace --check
> [ Violation detected ] release-notes/Cabal-3.8.1.0.md
> Error: Process completed with exit code 1.

I am using the vanilla `actions/checkout`: https://github.com/haskell/cabal/pull/8320/files#diff-0ea7b088f801fe875e2de46ed9180c6e04bb6f0bd9e7311428786195831e6b90R19

File: .github/workflows/whitespace.yml (hunk 2 above)
Comment (ulysses4ever) [reply .github/workflows/whitespace.yml:8]: All right, maybe it's fine.

**Comment by andreasabel** [comment]
> Would it make sense to add a line to CONTRIBUTING explaining how to run the check locally?

Just pushed a commit to that extend.

**Comment by Mikolaj** [comment]
> > Would it make sense to add a line to CONTRIBUTING explaining how to run the check locally?
> 
> Just pushed a commit to that extend.

Looks fine, thank you.

File: .github/workflows/whitespace.yml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    steps:
+    - uses: actions/checkout@v2
+      name: Checkout sources
+
+    - name: Create directory for binary
+      run: |
+        mkdir -p $HOME/.local/bin
+
+    - uses: actions/cache@v2
```
Comment (chshersh) [inline .github/workflows/whitespace.yml:26]: The latest version of the `checkout` and `cache` actions is `v3` so you can update it already 🙂

**Review by chshersh — APPROVED** [review]
The `fix-whitespace` tool is cool, I might want to use it in my own projects as well 🙂

File: .github/workflows/whitespace.yml (hunk 4 above)
Comment (andreasabel) [reply .github/workflows/whitespace.yml]: Done!

**Comment by andreasabel** [comment]
I reorganized the 6 commits into 3, but these 3 are independent and should be kept separate.

Why didn't label "merge me" do anything in 4 hours?

**Comment by ulysses4ever** [comment]
Because of #8285.

**Comment by andreasabel** [comment]
Sorry, I skipped the 2-day waiting period because this PR would rot too fast: it would get broken every time someone introduces a whitespace violation into the code base.

**Comment by ulysses4ever** [comment]
tldr; We need to backport this to 3.8 if we want a working CI on 3.8

Reasoning: see #8503 and #8505. They are required for working CI since ghcup updated the recommended GHC -- no matter the branch. And #8505 mentions the whitespace job, so it couldn't be (easily) backported without this one being backported first.

**Comment by ulysses4ever** [comment]
@Mergify backport 3.8
